savant/gstreamer/codecs.py

```diff
@@ -25,63 +25,105 @@ class CodecInfo:
     parser: Optional[str] = None
     """Gstreamer parser element."""
 
-    encoder_elements: Optional[List[str]] = None
-    """Gstreamer encoder elements.
-    Savant will use the first available element.
+    nv_encoder: Optional[str] = None
+    """Nvenc gstreamer encoder element.
+    Savant will use the it when encoder type is not specified.
+    """
+
+    sw_encoder: Optional[str] = None
+    """Software gstreamer encoder element.
+    Savant will use the it when encoder type is not specified and codec 
+    does not have NvEnc encoder.
     """
 
     @property
     def caps_with_params(self) -> str:
         """Caps with caps params string."""
         return ','.join([self.caps_name] + self.caps_params)
 
-    @property
-    def encoder(self) -> Optional[str]:
-        if not hasattr(self, '_encoder'):
-            if not self.encoder_elements:
-                encoder = None
-            elif len(self.encoder_elements) == 1:
-                encoder = self.encoder_elements[0]
+    def encoder(self, encoder_type: Optional[str]) -> Optional[str]:
+        """Get Gstreamer encoder element name.
+
+        :param encoder_type: Encoder type. Can be 'nvenc' or 'software'.
+                             When not specified nv_encoder will be used if codec has it
+                             or sw_encoder when codec does not have nv_encoder.
+        """
+
+        if not hasattr(self, '_nv_encoder'):
+            if self.nv_encoder is not None and _check_element_exists(self.nv_encoder):
+                self._nv_encoder = self.nv_encoder
+            else:
+                self._nv_encoder = None
+
+        if not hasattr(self, '_sw_encoder'):
+            if self.sw_encoder is not None and _check_element_exists(self.sw_encoder):
+                self._sw_encoder = self.sw_encoder
             else:
-                for element_name in self.encoder_elements:
-                    logger.debug('Check if element %r exists', element_name)
-                    elem_factory = Gst.ElementFactory.find(element_name)
-                    if elem_factory is not None:
-                        logger.debug('Found element %r', element_name)
-                        encoder = element_name
-                        break
-                    logger.debug('Element %r not found', element_name)
-                else:
-                    encoder = self.encoder_elements[-1]
+                self._sw_encoder = None
 
-            self._encoder = encoder
+        if encoder_type is None:
+            return self._nv_encoder or self._sw_encoder
 
-        return self._encoder
+        if encoder_type == 'nvenc':
```

Comment:
Why is it called nvenc? Can we call it hardware? I think it's easier to understand and if we suddenly support some other hardware, we can choose based on the hardware instead of the user being forced to choose.

Reply:
@dorgun won't fix now, let it be 'nvenc'.